---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

Check how the actual data looks like
```{r}
setwd("~/Box Sync/Danko_lab_work/mouse_AlleleDB/hmm/simulation/")
print("GM12878")
f1p="GM12878_GROseq_counts_hmm.txt"
f1=read.table(f1p,header=T)
median(f1$total_reads_count) 
dim(f1)
hist(f1$total_reads_count[f1$total_reads_count<100], main="GM12878")

print("CAST x B6")
f1p="CAST_B6_LZ_counts_hmm.txt"
f1=read.table(f1p,header=T)
median(f1$total_reads_count) 
dim(f1)
hist(f1$total_reads_count[f1$total_reads_count<100], main="CAST x B6")

```
Simulation of one HMM block
```{r}
#Simulation of one HMM block
#BlockLength=20
#ExpressionLevel=3
#MatBinoP=0.1
sim_HMM_block <- function(BlockLength, ExpressionLevel, MatBinoP){
  total_reads=rpois(BlockLength, ExpressionLevel)
  mat_reads=rbinom(BlockLength, total_reads, MatBinoP)
  SimState = rep(ifelse(MatBinoP<0.49, "P", ifelse(MatBinoP>0.51,"M", "S")),BlockLength)
  MatP= rep(round(MatBinoP, digits = 3),BlockLength)
  block=data.frame(total_reads, mat_reads,MatBinoP=MatP,SimState)
  return(block)
}
#b1=sim_HMM_block(10,5, 0.2)
#b1
```
Human
Average gene size in human is 10-15 Kb. If every SNP per Kb, then 10-15 SNPs per gene.

Mouse
http://www.informatics.jax.org/silver/frames/frame5-1.shtml
Nevertheless, a survey spanning all of the hundreds of mammalian genes characterized to date would seem to suggest that mdx is an extreme example, with most genes falling into the range of 10-80 kb, with an median size in the range of 20 kb. If every SNPs 160bp, then 125 SNPs per gene.
```{r}
# Simulation of 3 HMM blocks (Sym-Mat-Sym)
numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)
i=1
blockList=data.frame(sim_HMM_block(l[i],e[i], mat_p[i]), blockID=i)
for (i in 2:numberOfBlock){
  #print(i)
  #print(e[i])
  #print(mat_p[i])
  blockList= rbind.data.frame(blockList, cbind(sim_HMM_block(l[i],e[i], mat_p[i]),blockID=i))
}
blockList=blockList[blockList$total_reads>0,c("blockID", "SimState", "MatBinoP", "total_reads", "mat_reads")]
write.table(blockList, file = "similated_Counts.txt", quote = F, sep = "\t")
blockList
```

Simulation of a string of HMM blocks
```{r}
# Simulation of a string of HMM blocks
numberOfBlock=10
#legnth of each block
l <- sample(10:100, numberOfBlock, replace=T)
# expression level of each block
e <- sample(1:20, numberOfBlock, replace=T)
# mat p value of each block
mat_p <-runif(numberOfBlock, min=0, max=1)
i=1
blockList=data.frame(sim_HMM_block(l[i],e[i], mat_p[i]), blockID=i)
for (i in 2:numberOfBlock){
  print(i)
  #print(e[i])
  #print(mat_p[i])
  blockList= rbind.data.frame(blockList, cbind(sim_HMM_block(l[i],e[i], mat_p[i]),blockID=i))
}
blockList=blockList[blockList$total_reads>0,c("blockID", "SimState", "MatBinoP", "total_reads", "mat_reads")]
#which(names(blockList)=="blockID")
#View(blockList)
write.table(blockList, file = "similated_Counts.txt", quote = F, sep = "\t")
blockList
```


naive model:treating SNPs independently
```{r}
# naive model:treating SNPs independently.
blockList$AlleleDB_p_value <- 1
i=1
for (i in 1:dim(blockList)[1]){
  #if (blockList$total_reads[i] > 0){ #simulation with Poisoon generated 0 count
    a = binom.test(blockList$mat_reads[i], blockList$total_reads[i], 0.5)
    #print (a$p.value)
    blockList$AlleleDB_p_value[i] <- a$p.value    
  #}
}
#View(blockList)

p_threshold=0.05
blockList$AlleleDB_p_value_sig <- blockList$AlleleDB_p_value < p_threshold
blockList
```

69
p_Y_l  -188.887959957
p_Y_l  -191.107214928
A :  -219.04410696
secs:  -219.653127909
[[  9.99990000e-01   5.00000000e-06   5.00000000e-06]
 [  1.84668156e-03   9.96232518e-01   1.92080089e-03]
 [  5.00000000e-06   5.00000000e-06   9.99990000e-01]] 
 [0.89955685995989998, 0.50242206595539318, 0.10013546869624733] 
 -3765652.05209


python code to predict the state from HMM
HMM_prediction.py
```{python}
# python code to predict the state from HMM
# HMM_prediction.py

import numpy as np
from math import *
import scipy.stats
from sys import argv

###structure of hmm

##intial prob
I_s=0.5
I_m=0.25
I_p=0.25

##transition
# 0,1,2 = M, S, P

# emmision
binomlogpmf_dic={}
def get_emission_log_prob(x,n,p):
    if (x,n,p) not in binomlogpmf_dic:
        binomlogpmf_dic[(x,n,p)] = binomlogpmf(x, n, p)
    return binomlogpmf_dic[(x,n,p)]


def binomlogpmf(k, n, p):
    return scipy.stats.binom.logpmf(k, n, p)


def get_max_argmax(l):
    m = max(l)
    a = l.index(m)
    return m, a


binomtest_dic={}
def binomtest(x, n, p):
    x = min (x, n-x)
    if (x,n,p) not in binomtest_dic:
        binomtest_dic[(x,n,p)] = scipy.stats.binom_test(x, n, p)
    return binomtest_dic[(x,n,p)]


def viterbi (p, T, x, n):
    v = np.full((3, len(x)), float('-inf'))
    b = np.full((3, len(x)), int(3), dtype=int)
    #Initialization
    v[0, 0] = log(I_m) + get_emission_log_prob(x[0],n[0],p[0])
    v[1, 0] = log(I_s) + get_emission_log_prob(x[0],n[0],p[1])
    v[2, 0] = log(I_p) + get_emission_log_prob(x[0],n[0],p[2])
    b[0 ,0] = 0
    b[1 ,0] = 1
    b[2 ,0] = 2
    #Iteration
    for i in xrange(1, len(x)):
        v[0, i], b[0, i] = get_max_argmax([v[0, i-1] + T[0,0],  v[1, i-1] + T[1,0], v[2, i-1] + T[2,0]])
        v[1, i], b[1, i] = get_max_argmax([v[0, i-1] + T[0,1],  v[1, i-1] + T[1,1], v[2, i-1] + T[2,1]])
        v[2, i], b[2, i] = get_max_argmax([v[0, i-1] + T[0,2],  v[1, i-1] + T[1,2], v[2, i-1] + T[2,2]])
        v[0, i] += get_emission_log_prob(x[i],n[i],p[0])
        v[1, i] += get_emission_log_prob(x[i],n[i],p[1])
        v[2, i] += get_emission_log_prob(x[i],n[i],p[2])
    #track back pointer
    viterbi_path_backward=[]
    i = len(x) - 1
    f = np.argmax(v[:,i])
    viterbi_path_backward.append(f)
    while (i > 0):
        f = b[f,i]
        viterbi_path_backward.append(f)
        i -= 1
    
    viterbi_path_forward = np.array(viterbi_path_backward[::-1])
    return viterbi_path_forward


### input data for viterbi
# seperate the autosome
# seperate plus and minus strand
#def hmm_prediction(f_v, strand, t,new_T, new_P):
if len(argv)==1:
    f_v = "similated_Counts.txt"
else:
    f_v=argv[1]
data_v = np.loadtxt(f_v, dtype=str ,delimiter='\t', usecols=range(0,6), skiprows=1)
mat_v  = np.loadtxt(f_v, dtype=int ,delimiter='\t', usecols=[5], skiprows=1)
total_v = np.loadtxt(f_v, dtype=int ,delimiter='\t', usecols=[4], skiprows=1)
t = 1e-05
t_mm, t_ms, t_mp = 1-t, t/2, t/2
t_sm, t_ss, t_sp =  1.84668156e-03, 9.96232518e-01, 1.92080089e-03
t_pm, t_ps, t_pp = t/2, t/2, 1-t
T = np.log(np.array( [[t_mm, t_ms, t_mp],[t_sm, t_ss, t_sp],[t_pm, t_ps, t_pp]]))


p_m, p_s,p_p = 0.9, 0.5, 0.1
p = [p_m, p_s,p_p]
#p= [0.89955685995989998, 0.50242206595539318, 0.10013546869624733] 
v_path=viterbi (p, T,x=mat_v, n=total_v)


state_map={0:"M", 1:"S", 2:"P"}
with open(f_v[0:-4]+'_AlleleHMM.txt', 'w') as out:
    out.write(open(f_v).readlines()[0].strip())
    out.write('\thmm_state\n')
    for i in xrange(data_v.shape[0]):
        out.write('\t'.join(data_v[i]))
        out.write('\t'+state_map[v_path[i]])
        out.write('\n')

```


```{r}
# combine the reads in the same HMM blocks and perform bionomial test
hmm_result=read.table("similated_Counts_AlleleHMM.txt",header=T)
#View(hmm_result)
# number of switch between M and P
hmm_result$tag=0
hmm_result$accum_total_reads=0
hmm_result$accum_mat_reads=0
hmm_result$p_value = 1
i=1
accum_total_reads=hmm_result$total_reads[i]
accum_mat_reads=hmm_result$mat_reads[i]
hmm_result$tag[i]=1
for (i in 2:dim(hmm_result)[1]){
  if (hmm_result$hmm_state[i] == hmm_result$hmm_state[i-1]){
   hmm_result$tag[i]=1
   accum_total_reads <- accum_total_reads + hmm_result$total_reads[i]
   accum_mat_reads <- accum_mat_reads + hmm_result$mat_reads[i]
  }
  else{
    hmm_result$accum_total_reads[hmm_result$tag==1] <- accum_total_reads
    hmm_result$accum_mat_reads[hmm_result$tag==1] <- accum_mat_reads
    a = binom.test(accum_mat_reads, accum_total_reads, 0.5)
    hmm_result$p_value[hmm_result$tag==1] <- a$p.value
    
    hmm_result$tag=0
    hmm_result$tag[i]=1
    accum_total_reads <- hmm_result$total_reads[i]
    accum_mat_reads   <- hmm_result$mat_reads[i]
  }
}
hmm_result$accum_total_reads[hmm_result$tag==1] <- accum_total_reads
hmm_result$accum_mat_reads[hmm_result$tag==1] <- accum_mat_reads
a = binom.test(accum_mat_reads, accum_total_reads, 0.5)
hmm_result$p_value[hmm_result$tag==1] <- a$p.value
View(hmm_result)
blockList$AlleleHMM_p_value_sig <- hmm_result$p_value < p_threshold
blockList$HMM_p_value <- hmm_result$p_value
blockList$HMM_accum_total_reads <- hmm_result$accum_total_reads
blockList$HMM_accum_mat_reads<- hmm_result$accum_mat_reads
blockList$HMM_state<- hmm_result$hmm_state
```

Sensitivity  test (remove later?)
```{r}
# MatBionP<0.3 or >0.7 and significant
dim(blockList)
sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7)
sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7 & blockList$AlleleDB_p_value <0.05)
sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7 & blockList$HMM_p_value <0.05)
print(paste("Sensitivity for AlleleDB is",sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7 & blockList$AlleleDB_p_value <0.05) /sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7)))
 
print(paste("Sensitivity for AlleleHMM is",sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7 & blockList$HMM_p_value <0.05) /sum(blockList$MatBinoP<0.3 | blockList$MatBinoP >0.7)))
 

```

wrap codes into a function to test sensitivity 
```{r}
## wrap into a function to test sensitivity 

#Simulation of one HMM block
#BlockLength=20
#ExpressionLevel=3
#MatBinoP=0.1
sim_HMM_block <- function(BlockLength, ExpressionLevel, MatBinoP){
  total_reads=rpois(BlockLength, ExpressionLevel)
  mat_reads=rbinom(BlockLength, total_reads, MatBinoP)
  SimState = rep(ifelse(MatBinoP<0.49, "P", ifelse(MatBinoP>0.51,"M", "S")),BlockLength)
  MatP= rep(round(MatBinoP, digits = 3),BlockLength)
  block=data.frame(total_reads, mat_reads,MatBinoP=MatP,SimState)
  return(block)
}
#b1=sim_HMM_block(10,5, 0.2)
#b1


numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

Sensitivity<- function(l, e, mat_p){
  # Simulation of 3 HMM blocks (Sym-Mat-Sym)
  i=1
  blockList=data.frame(sim_HMM_block(l[i],e[i], mat_p[i]), blockID=i)
  for (i in 2:numberOfBlock){
    blockList= rbind.data.frame(blockList, cbind(sim_HMM_block(l[i],e[i], mat_p[i]),blockID=i))
  }
  blockList=blockList[blockList$total_reads>0,c("blockID", "SimState", "MatBinoP", "total_reads",  "mat_reads")]
  #b_tmp=tempfile(tmpdir ="/workdir/sc2457/HMM_simulation/toremove",fileext = ".txt")
  b_tmp=tempfile(fileext = ".txt")
  write.table(blockList, file = b_tmp, quote = F, sep = "\t")
  #View(blockList)
  
  # naive model:treating SNPs independently.
  blockList$AlleleDB_p_value <- 1
  i=1
  for (i in 1:dim(blockList)[1]){
    a = binom.test(blockList$mat_reads[i], blockList$total_reads[i], 0.5)
    blockList$AlleleDB_p_value[i] <- a$p.value    
  }

 ## HMM prediction
  system(paste("python HMM_prediction.py",b_tmp), wait=TRUE)
 ## HMM binomial test
 
 # combine the reads in the same HMM blocks and perform bionomial test
 hmm_result=read.table(paste(substr(b_tmp,1,nchar(b_tmp)-4),"_AlleleHMM.txt",sep = ""),header=T)
 hmm_result$tag=0
 hmm_result$accum_total_reads=0
 hmm_result$accum_mat_reads=0
 hmm_result$p_value = 1
 i=1
 accum_total_reads=hmm_result$total_reads[i]
 accum_mat_reads=hmm_result$mat_reads[i]
 hmm_result$tag[i]=1
 for (i in 2:dim(hmm_result)[1]){
   if (hmm_result$hmm_state[i] == hmm_result$hmm_state[i-1]){
     hmm_result$tag[i]=1
     accum_total_reads <- accum_total_reads + hmm_result$total_reads[i]
     accum_mat_reads <- accum_mat_reads + hmm_result$mat_reads[i]
   }
   else{
     hmm_result$accum_total_reads[hmm_result$tag==1] <- accum_total_reads
     hmm_result$accum_mat_reads[hmm_result$tag==1] <- accum_mat_reads
     a = binom.test(accum_mat_reads, accum_total_reads, 0.5)
     hmm_result$p_value[hmm_result$tag==1] <- a$p.value
     
     hmm_result$tag=0
     hmm_result$tag[i]=1
     accum_total_reads <- hmm_result$total_reads[i]
     accum_mat_reads   <- hmm_result$mat_reads[i]
   }
 }
 hmm_result$accum_total_reads[hmm_result$tag==1] <- accum_total_reads
 hmm_result$accum_mat_reads[hmm_result$tag==1] <- accum_mat_reads
 a = binom.test(accum_mat_reads, accum_total_reads, 0.5)
 hmm_result$p_value[hmm_result$tag==1] <- a$p.value
 #print(dim(blockList)[1] == dim(hmm_result)[1])
 
 blockList$HMM_p_value <- hmm_result$p_value
 blockList$HMM_accum_total_reads <- hmm_result$accum_total_reads
 blockList$HMM_accum_mat_reads<- hmm_result$accum_mat_reads
 blockList$HMM_state<- hmm_result$hmm_state
 ##test Sensitivity
 if (sum(blockList$SimState!="S") !=0) {
   SNP_Sensitivity = sum(blockList$AlleleDB_p_value <0.05 & blockList$SimState!="S") /sum(blockList$SimState!="S") 
   AlleleHMM_Sensitivity = sum(blockList$HMM_p_value <0.05 & blockList$SimState!="S") /sum(blockList$SimState!="S") 
 }  else{
    SNP_Sensitivity = sum(blockList$AlleleDB_p_value <0.05 & blockList$SimState!="S") /(sum(blockList$SimState!="S")+1) 
   AlleleHMM_Sensitivity = sum(blockList$HMM_p_value <0.05 & blockList$SimState!="S") /(sum(blockList$SimState!="S")+1) 
 }
 
 # specificity
    SNP_Specificity = sum(blockList$AlleleDB_p_value >=0.05 & blockList$SimState=="S") /sum(blockList$SimState=="S") 
   AlleleHMM_Specificity = sum(blockList$HMM_state=="S" & blockList$SimState=="S") /sum(blockList$SimState=="S") 
 #print(AlleleHMM_Specificity)
   
  return (list(SNP_Sensitivity=SNP_Sensitivity, 
               AlleleHMM_Sensitivity=AlleleHMM_Sensitivity,
               SNP_Specificity = SNP_Specificity,
               AlleleHMM_Specificity = AlleleHMM_Specificity
               #, data=blockList 
               ))
}

Sensitivity_iter<- function(iteration,l, e, mat_p){
  SNP_sen_list=c()
  HMM_sen_list=c()
  SNP_spec_list=c()
  HMM_spec_list=c()
  for (i in 1:iteration) {
    #print (i)
    s=Sensitivity(l, e, mat_p)
    SNP_sen_list=c(SNP_sen_list, s$SNP_Sensitivity)
    HMM_sen_list=c(HMM_sen_list, s$AlleleHMM_Sensitivity)
    SNP_spec_list=c(SNP_spec_list, s$SNP_Specificity)
    HMM_spec_list=c(HMM_spec_list, s$AlleleHMM_Specificity)
  }
  return (c(mean(SNP_sen_list),mean(HMM_sen_list),
            sd(SNP_sen_list)/sqrt(length(SNP_sen_list)),
            sd(HMM_sen_list)/sqrt(length(HMM_sen_list)),
            mean(SNP_spec_list),mean(HMM_spec_list),
            sd(SNP_spec_list)/sqrt(length(SNP_spec_list)),
            sd(HMM_spec_list)/sqrt(length(HMM_spec_list))))
  #return (c(mean(SNP_sen_list), mean(HMM_sen_list)))
}


```


Mat P and sensitivity
```{r}
numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

iteration=3
#mat_p_list=c()
#SNP_sen_list=c()
#HMM_sen_list=c()
#SNP_sen_se_list=c()
#HMM_sen_se_list=c()
result=c()
for (mat_p_2 in seq(0.1,0.9,0.05)){
  mat_p_list = c(mat_p_list, mat_p_2) 
  mat_p=c(0.5,mat_p_2,0.5)
  #print(mat_p)
  #aveS = Sensitivity_iter(iteration,l, e, mat_p)
  #SNP_sen_list=c(SNP_sen_list, aveS[1])
  #HMM_sen_list=c(HMM_sen_list, aveS[2])
  #SNP_sen_se_list=c(SNP_sen_se_list, aveS[3])
  #HMM_sen_se_list=c(HMM_sen_se_list, aveS[4])
  result=c(result,  mat_p_2, Sensitivity_iter(iteration,l, e, mat_p))
  }
View(result)

m1=data.frame(result[seq(1,length(result),9)],
              result[seq(2,length(result),9)], 
              result[seq(4,length(result),9)],
              result[seq(6,length(result),9)],
              result[seq(8,length(result),9)])
names(m1)=c("matP","senMean", "senSE", "specMean", "specSE")
m1$method="SNP independantly"
#View(m)
m2=data.frame(result[seq(1,length(result),9)],
              result[seq(3,length(result),9)], 
              result[seq(5,length(result),9)],
              result[seq(7,length(result),9)],
              result[seq(9,length(result),9)])
names(m2)=c("matP","senMean", "senSE","specMean", "specSE")
m2$method="Allele HMM"
m=rbind.data.frame(m1,m2)
write.table(m, file = "matP_and_sensitivity.txt", quote = F, sep = "\t",row.names = F)


#m1=data.frame(mat_p_list,SNP_sen_list,SNP_sen_se_list)
#names(m1)=c("matP","senMean", "senSE")
#m1$method="SNP independantly"
#View(m)
#m2=data.frame(mat_p_list,HMM_sen_list, HMM_sen_se_list)
#names(m2)=c("matP","senMean", "senSE")
#m2$method="Allele HMM"
#m=rbind.data.frame(m1,m2)
#write.table(m, file = "matP_and_sensitivity.txt", quote = F, sep = "\t",row.names = F)

#View(m)
library(ggplot2)
# Standard error of the mean
#pd <- position_dodge(0)
pdf("matP_and_sensitivity.pdf")
ggplot(m, aes(x=matP, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.01) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()

pdf("matP_and_specificity.pdf")
ggplot(m, aes(x=matP, y=specMean, colour=method)) + 
    geom_errorbar(aes(ymin=specMean-specSE, ymax=specMean+specSE), width=.01) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()

plot(mat_p_list, HMM_sen_list, main="mat Binomial P", ylab="sensitivity", xlab="mat P", type="l", col="dark orange",lwd=2)
lines(mat_p_list, SNP_sen_list, col="blue",lwd=2)  
legend("bottomleft", 
       legend = c("Allele HMM", "SNP independantly"),
       #pch=15,
       cex=1, 
       lty=1, lwd=2,
       col=c("dark orange","blue"), bty = "n")
#http://127.0.0.1:40576/chunk_output/B1844C7ED108D0A9/3235B715/cg1kuvtz4y6np/000081.png
```

test multi-threads with MatP
```{r}
## test multi-threads with MatP

numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

library(parallel)
m_test <- function(mat_p_2, iteration){
  #mat_p_list = c(mat_p_list, mat_p_2) 
  mat_p=c(0.5,mat_p_2,0.5)
  #print(mat_p)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(mat_p_2,aveS))
}

test=mclapply(seq(0.1,0.9,0.05),FUN=function(idx){m_test(idx,1000)}, mc.cores = 1)

result=unlist(test)

m1=data.frame(result[seq(1,length(result),9)],
              result[seq(2,length(result),9)], 
              result[seq(4,length(result),9)],
              result[seq(6,length(result),9)],
              result[seq(8,length(result),9)])
names(m1)=c("matP","senMean", "senSE", "specMean", "specSE")
m1$method="SNP independantly"
#View(m)
m2=data.frame(result[seq(1,length(result),9)],
              result[seq(3,length(result),9)], 
              result[seq(5,length(result),9)],
              result[seq(7,length(result),9)],
              result[seq(9,length(result),9)])
names(m2)=c("matP","senMean", "senSE","specMean", "specSE")
m2$method="Allele HMM"
m=rbind.data.frame(m1,m2)
write.table(m, file = "matP_and_sensitivity_and_specificity.txt", quote = F, sep = "\t",row.names = F)

m_o = read.table("matP_and_sensitivity.txt", sep = "\t",header = T)
View(m_o)
library(ggplot2)
pdf("matP_and_sensitivity.pdf")
ggplot(m, aes(x=matP, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.01) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()
pdf("matP_and_specificity.pdf")
ggplot(m, aes(x=matP, y=specMean, colour=method)) + 
    geom_errorbar(aes(ymin=specMean-specSE, ymax=specMean+specSE), width=.01) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()

plot(t[seq(1,length(t),9)],t[seq(3,length(t),9)], main="mat Binomial P", ylab="sensitivity", xlab="mat P", type="l", col="dark orange",lwd=2)
lines(t[seq(1,length(t),9)],t[seq(2,length(t),9)], col="blue",lwd=2)  
legend("bottomleft", 
       legend = c("Allele HMM", "SNP independantly"),
       #pch=15,
       cex=1, 
       lty=1, lwd=2,
       col=c("dark orange","blue"), bty = "n")

```

test multi-threads with expression level
```{r}
## test multi-threads with expression level
numberOfBlock <- 3
#legnth of each block
l <- c(10,10,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

library(parallel)
e_test <- function(e_2, iteration){
  e=c(10, e_2, 10)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(e_2,aveS))
}
e_test_out=mclapply(seq(1,50,1),FUN=function(idx){e_test(idx,3)}, mc.cores = 1)






t=unlist(e_test_out)
e1=data.frame(t[seq(1,length(t),9)],t[seq(2,length(t),9)],t[seq(4,length(t),9)])
names(e1)=c("readCount","senMean", "senSE")
e1$method="SNP independantly"
e2=data.frame(t[seq(1,length(t),9)],t[seq(3,length(t),9)],t[seq(5,length(t),9)])
names(e2)=c("readCount","senMean", "senSE")
e2$method="Allele HMM"
e=rbind.data.frame(e1,e2)
write.table(e, file = "exp_and_sensitivity.txt", quote = F, sep = "\t",row.names = F)

library(ggplot2)
pdf("exp_and_sensitivity.pdf")
ggplot(e, aes(x=readCount, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.01) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()

```


Expression level and sensitivity
```{r}
#Expression level and sensitivity

numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

iteration=1000
e_list=c()
e_SNP_sen_list=c()
e_HMM_sen_list=c()
e_SNP_sen_se_list=c()
e_HMM_sen_se_list=c()
for (e_2 in seq(1,50,1)){
  e_list = c(e_list, e_2) 
  e=c(10, e_2, 10)
  #print(mat_p)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  e_SNP_sen_list=c(e_SNP_sen_list, aveS[1])
  e_HMM_sen_list=c(e_HMM_sen_list, aveS[2])
  e_SNP_sen_se_list=c(e_SNP_sen_se_list, aveS[3])
  e_HMM_sen_se_list=c(e_HMM_sen_se_list, aveS[4])
}

e1=data.frame(e_list,e_SNP_sen_list,e_SNP_sen_se_list)
names(e1)=c("readCount","senMean", "senSE")
e1$method="SNP independantly"
e2=data.frame(e_list,e_HMM_sen_list, e_HMM_sen_se_list)
names(e2)=c("readCount","senMean", "senSE")
e2$method="Allele HMM"
e=rbind.data.frame(e1,e2)
#View(e)
write.table(e, file = "exp_and_sensitivity.txt", quote = F, sep = "\t",row.names = F)


#View(e)
library(ggplot2)
# Standard error of the mean
#pd <- position_dodge(0)
pdf("exp_and_sensitivity.pdf")
ggplot(e, aes(x=readCount, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.01) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()



plot(e_list, e_HMM_sen_list, 
     ylim = c(min(min(e_HMM_sen_list),min(e_SNP_sen_list)), 1), lwd=2,
     ylab="sensitivity", xlab="expression level (mapped reads per SNP)", 
     type="l", col="dark orange")
lines(e_list, e_SNP_sen_list, lwd=2, col="blue")  
legend("bottomright", 
       legend = c("Allele HMM", "SNP independantly"),
       #pch=15,
       cex=1, 
       lty=1, lwd=2,
       col=c("dark orange","blue"), bty = "n")

```

test multi-threads with block length
```{r}
## test multi-threads with block length
numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

library(parallel)
l_test <- function(l_2, iteration){
  l=c(10, l_2, 10)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(l_2,aveS))
}
l_test_out=mclapply(seq(1,20,1),FUN=function(idx){l_test(idx,3)}, mc.cores = 3)
```

Block Length and sensitivity
```{r}
#Block Length and sensitivity

numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.9,0.5)

iteration=1000
l_list=c()
l_SNP_sen_list=c()
l_HMM_sen_list=c()
l_SNP_sen_se_list=c()
l_HMM_sen_se_list=c()
for (l_2 in seq(1,50,1)){
  l_list = c(l_list, l_2) 
  l=c(10, l_2, 10)
  #print(mat_p)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  l_SNP_sen_list=c(l_SNP_sen_list, aveS[1])
  l_HMM_sen_list=c(l_HMM_sen_list, aveS[2])
  l_SNP_sen_se_list=c(l_SNP_sen_se_list, aveS[3])
  l_HMM_sen_se_list=c(l_HMM_sen_se_list, aveS[4])
}

l1=data.frame(l_list,l_SNP_sen_list,l_SNP_sen_se_list)
names(l1)=c("length","senMean", "senSE")
l1$method="SNP independantly"
l2=data.frame(l_list,l_HMM_sen_list, l_HMM_sen_se_list)
names(l2)=c("length","senMean", "senSE")
l2$method="Allele HMM"
l=rbind.data.frame(l1,l2)
#View(e)
write.table(l, file = "length_and_sensitivity.txt", quote = F, sep = "\t",row.names = F)

l = read.table("length_and_sensitivity_iter1000.txt", sep = "\t",header = T)
#l=l[l$method=="SNP independantly",]


#View(l)
library(ggplot2)
# Standard error of the mean
#pd <- position_dodge(0)
pdf("length_and_sensitivity.pdf")
ggplot(l, aes(x=length, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.5) +
    geom_line() +
    geom_point() #+ 
    #theme_bw()
dev.off()


plot(l_list, l_HMM_sen_list, 
     ylim = c(min(min(l_HMM_sen_list),min(l_SNP_sen_list)), 1), lwd=2,
     ylab="sensitivity", xlab="Gene length", 
     type="l", col="dark orange")
lines(l_list, l_SNP_sen_list, lwd=2, col="blue")  
legend("bottomright", 
       legend = c("Allele HMM", "SNP independantly"),
       #pch=15,
       cex=1, 
       lty=1, lwd=2,
       col=c("dark orange","blue"), bty = "n")

```


specificity of SSS block
```{r}
#specificity of SSS block

# helper functions:
result_to_ggplot <- function(test_out,txt_name,test_para, pdf_name, y_lab ){
  result=unlist(test_out)  
  m1=data.frame(result[seq(1,length(result),9)],
                result[seq(2,length(result),9)], 
                result[seq(4,length(result),9)],
                result[seq(6,length(result),9)],
                result[seq(8,length(result),9)])
  names(m1)=c("test_para","senMean", "senSE", "specMean", "specSE")
  m1$method="SNP independantly"
  #View(m)
  m2=data.frame(result[seq(1,length(result),9)],
                result[seq(3,length(result),9)], 
                result[seq(5,length(result),9)],
                result[seq(7,length(result),9)],
                result[seq(9,length(result),9)])
  names(m2)=c("test_para","senMean", "senSE","specMean", "specSE")
  m2$method="Allele HMM"
  m=rbind.data.frame(m1,m2)
  #pdf_name="test.pdf"
  #ylab="specificity"
  #test_para="length"
  pdf(pdf_name)
  myplot<- ggplot(m, aes(x=test_para, y=specMean, colour=method)) + 
    geom_errorbar(aes(ymin=specMean-specSE, ymax=specMean+specSE), width=.01) +
    geom_line() +
    geom_point()+
    xlab(test_para)+
    ylab(ylab); 
  #theme_bw()
  print(myplot)
  dev.off()
  
  names(m)=c(test_para,"senMean", "senSE", "specMean", "specSE", "method")
  write.table(m, file = txt_name, quote = F, sep = "\t",row.names = F)
}

# specificity of SSS block
numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.5,0.5,0.5)

library(parallel)
l_test <- function(l_2, iteration){
  l=c(10, l_2, 10)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(l_2,aveS))
}

l_test_out=mclapply(seq(1,100,1),FUN=function(idx){l_test(idx,1000)}, mc.cores = 10)
result_to_ggplot(l_test_out,"length_and_specificity.txt","length", "length_and_specificity.pdf","Specificity")
result_to_ggplot_both(l_test_out,"length_SSS.txt","length_SSS") 

e_test <- function(e_2, iteration){
  e=c(10, e_2, 10)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(e_2,aveS))
}
library(parallel)
e_test_out=mclapply(seq(1,50,1),FUN=function(idx){e_test(idx,1000)}, mc.cores = 10)
result_to_ggplot(e_test_out,"expressionLevel_and_specificity.txt","expressionLevel", "expressionLevel_and_specificity.pdf","Specificity")
result_to_ggplot_both(e_test_out,"expressionLevel_SSS.txt","expressionLevel_SSS") 


```


test specificity when flank by M block
```{r}
# (Mat)-(Sym)-(Mat) and test how long/expression level the Sym block need to be to distinct apart 

# specificity of SSS block
numberOfBlock <- 3
#legnth of each block
l <- c(10,100,10)
# expression level of each block
e <- c(10,10,10)
mat_p <- c(0.9,0.5,0.9)

# helper functions:
result_to_ggplot_both <- function(test_out,txt_name,test_para){
  result=unlist(test_out)  
  m1=data.frame(result[seq(1,length(result),9)],
                result[seq(2,length(result),9)], 
                result[seq(4,length(result),9)],
                result[seq(6,length(result),9)],
                result[seq(8,length(result),9)])
  names(m1)=c("test_para","senMean", "senSE", "specMean", "specSE")
  m1$method="SNP independantly"
  #View(m)
  m2=data.frame(result[seq(1,length(result),9)],
                result[seq(3,length(result),9)], 
                result[seq(5,length(result),9)],
                result[seq(7,length(result),9)],
                result[seq(9,length(result),9)])
  names(m2)=c("test_para","senMean", "senSE","specMean", "specSE")
  m2$method="Allele HMM"
  m=rbind.data.frame(m1,m2)
  pdf(paste(test_para,"_Specificity.pdf",sep=""))
  myplot<- ggplot(m, aes(x=test_para, y=specMean, colour=method)) + 
    geom_errorbar(aes(ymin=specMean-specSE, ymax=specMean+specSE), width=.01) +
    geom_line() +
    geom_point()+
    xlab(test_para)+
    ylab("Specificity")#+  
    #ylim(c(0.96,1)) 
  #theme_bw()
  print(myplot+ scale_color_manual(values=c("red", "blue")))
  dev.off()
  
    pdf(paste(test_para,"_Sensitivity.pdf",sep=""))
  myplot<- ggplot(m, aes(x=test_para, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.01) +
    geom_line() +
    geom_point()+
    xlab(test_para)+
    ylab("Sensitivity")#+  
    #ylim(c(0,1)) 
  #theme_bw()
  
  print(myplot+ scale_color_manual(values=c("red", "blue")))
  dev.off()
  
  
  names(m)=c(test_para,"senMean", "senSE", "specMean", "specSE", "method")
  write.table(m, file = txt_name, quote = F, sep = "\t",row.names = F)
}

library(parallel)
l_test <- function(l_2, iteration){
  l=c(10, l_2, 10)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(l_2,aveS))
}

l_test_out=mclapply(seq(1,30,1),FUN=function(idx){l_test(idx,1000)}, mc.cores = 10)
#result_to_ggplot_both <- function(test_out,txt_name,test_para){
result_to_ggplot_both(l_test_out,"length_MSM.txt","length_MSM") 


e_test <- function(e_2, iteration){
  e=c(10, e_2, 10)
  aveS = Sensitivity_iter(iteration,l, e, mat_p)
  return(c(e_2,aveS))
}
library(parallel)
e_test_out=mclapply(seq(1,300,1),FUN=function(idx){e_test(idx,1000)}, mc.cores = 10)
#result_to_ggplot(e_test_out,"expressionLevel_and_specificity_MSM.txt","expressionLevel", "expressionLevel_and_specificity_MSM.pdf","Specificity")
result_to_ggplot_both(e_test_out,"expressionLevel_MSM.txt","expressionLevel_MSM") 

```


Code to make plots
```{r}
setwd("~/Box Sync/Danko_lab_work/mouse_AlleleDB/hmm/simulation")
m=read.table("SSS_blocks/expressionLevel_SSS.txt", header=T,sep = "\t")
mplot <- ggplot(m, aes(x=expressionLevel_SSS, y=specMean, colour=method)) + 
    geom_errorbar(aes(ymin=specMean-specSE, ymax=specMean+specSE), width=.1) +
    geom_line() +
    geom_point()+
    xlab("expression level")+
    ylab("Specificity")+
  ylim(c(0.96,1))+
  xlab("middle block expression level")
mplot + scale_color_manual(values=c("red", "blue"))
#output 8x5

m=read.table("SSS_blocks/length_SSS.txt", header=T,sep = "\t")
mplot <- ggplot(m, aes(x=length_SSS, y=specMean, colour=method)) + 
    geom_errorbar(aes(ymin=specMean-specSE, ymax=specMean+specSE), width=.1) +
    geom_line() +
    geom_point()+
    xlab("length")+
    ylab("Specificity")+
  ylim(c(0.96,1))+
  xlab("middle block length")
mplot + scale_color_manual(values=c("red", "blue"))
#output 8x5

m=read.table("matP_and_sensitivity_iter1000.txt", header=T,sep = "\t")
  mplot<- ggplot(m, aes(x=matP, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.03) +
    geom_line() +
    geom_point()+
    xlab("Mat Binomial P")+
    ylab("Sensitivity")+  
    ylim(c(0,1)) 
  #theme_bw()
  mplot + scale_color_manual(values=c("red", "blue"))

m=read.table("length_and_sensitivity_iter1000.txt", header=T,sep = "\t")
  mplot<- ggplot(m, aes(x=length, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.1) +
    geom_line() +
    geom_point()+
    xlab("middle block length")+
    ylab("Sensitivity")+  
    ylim(c(0,1))+
    xlim(c(0,20))
  #theme_bw()
  mplot + scale_color_manual(values=c("red", "blue"))

m=read.table("exp_and_sensitivity_iter1000.txt", header=T,sep = "\t")
  mplot<- ggplot(m, aes(x=readCount, y=senMean, colour=method)) + 
    geom_errorbar(aes(ymin=senMean-senSE, ymax=senMean+senSE), width=.1) +
    geom_line() +
    geom_point()+
    xlab("average read count per SNP")+
    ylab("Sensitivity")+  
    ylim(c(0,1))+
    xlim(c(0,40))
  #theme_bw()
  mplot + scale_color_manual(values=c("red", "blue"))
```




Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

